flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java

```diff
@@ -255,4 +257,34 @@ public static boolean shouldRollBack(
                 .minus(readinessTimeout)
                 .isAfter(Instant.ofEpochMilli(reconciliationStatus.getReconciliationTimestamp()));
     }
+
+    public static <
+                    SPEC extends AbstractFlinkSpec,
+                    STATUS extends CommonStatus<SPEC>,
+                    CR extends CustomResource<SPEC, STATUS>>
+            boolean applyValidationErrorAndResetSpec(
+                    CR deployment, CommonStatus<?> prevStatus, String newErr) {
+
+        if (!newErr.equals(prevStatus.getError())) {
+            LOG.error("Validation failed: " + newErr);
+        }
+
+        ReconciliationUtils.updateForReconciliationError(deployment, newErr);
+
+        var lastReconciledSpec =
+                deployment.getStatus().getReconciliationStatus().deserializeLastReconciledSpec();
+        if (lastReconciledSpec == null) {
+            // Validation failed before anything was deployed, nothing to do
+            return false;
+        } else {
+            // We need to observe/reconcile using the last version of the deployment spec
+            deployment.setSpec(lastReconciledSpec);
```

Comment:
This means we will change the spec when it generates the validation error, will this break the principle that not change the spec directly ?

Reply:
I have a doubt that can we enforce user to use the operator with the webhook? I think it will make the things much easier, because it will not apply the unexpected spec succeed. Do you know whether there is strong reason for users to disable the webhook?

Reply:
The spec is never changed. We always change it back to the original one and we only update the status.

that’s the whole idea :)

Reply:
I wish we could always use the webhook but even in our environments it’s not always possible due to access limitations.

Reply:
Get it